src/ShaderGen.Tests/AutoGenerated/TestSet.cs

```diff
@@ -0,0 +1,230 @@
+﻿using System;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Threading;
+using ShaderGen.Tests.Tools;
+using Veldrid;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace ShaderGen.Tests.AutoGenerated
+{
+    /// <summary>
+    /// An individual test set.
+    /// </summary>
+    internal class TestSet
+    {
+        /// <summary>
+        /// The owner <see cref="TestSets"/> collection.
+        /// </summary>
+        public readonly TestSets TestSets;
+
+        /// <summary>
+        /// The test name.
+        /// </summary>
+        public readonly string Name;
+
+        /// <summary>
+        /// The <see cref="ToolChain"/> if any; otherwise <see langword="null"/>.
+        /// </summary>
+        public readonly ToolChain ToolChain;
+
+        /// <summary>
+        /// The <see cref="LanguageBackend"/> if any; otherwise <see langword="null"/>.
+        /// </summary>
+        public readonly LanguageBackend Backend;
+
+        /// <summary>
+        /// The results data.
+        /// </summary>
+        public byte[] Results { get; private set; }
+
+        private int _executed = 0;
+        /// <summary>
+        /// Gets a value indicating whether this <see cref="TestSet"/> has been executed yet.
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if executed; otherwise, <c>false</c>.
+        /// </value>
+        public bool Executed => _executed > 0;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TestSet"/> class.
+        /// </summary>
+        /// <param name="testSets">The test sets.</param>
+        /// <param name="toolChain">The tool chain (if any).</param>
+        public TestSet(TestSets testSets, ToolChain toolChain = null)
+        {
+            TestSets = testSets;
+            Name = toolChain?.GraphicsBackend.ToString() ?? "CPU";
+            ToolChain = toolChain;
+            Backend = toolChain?.CreateBackend(testSets.Compilation);
+            Results = Array.Empty<byte>();
+        }
+
+        /// <summary>
+        /// Executes the specified test.
+        /// </summary>
+        /// <param name="generationResult">The generation result.</param>
+        /// <param name="csFunctionName">Name of the cs function.</param>
+        /// <param name="output">The output.</param>
+        public void Execute(
+            ShaderGenerationResult generationResult,
+            string csFunctionName,
+            ITestOutputHelper output)
+        {
+            if (Interlocked.CompareExchange(ref _executed, 1, 0) > 0)
```

Comment:
I don't see why this is necessary. It doesn't look like the tests are executed in parallel, or even more than once anyways. Is there something I'm missing?

Reply:
Nope, soz, force of habit, most of the code I write is multi-threaded, so I'm just used to implementing barriers using this pattern.